Zend/zend_gc.c

```diff
@@ -184,47 +245,56 @@ static zend_gc_globals gc_globals;
 # define GC_TRACE(str)
 #endif
 
-static zend_always_inline void gc_ref_set_info(zend_refcounted *ref, uint32_t info)
+static zend_always_inline uint32_t gc_compress(uint32_t idx)
 {
-	GC_TYPE_INFO(ref) = (info << GC_INFO_SHIFT)
-		| (GC_TYPE_INFO(ref) & (GC_TYPE_MASK | GC_FLAGS_MASK));
+	ZEND_ASSERT(GC_NEED_COMPRESSION(idx));
+	return GC_MAX_UNCOMPRESSED + ((idx - GC_MAX_UNCOMPRESSED) % GC_COMPRESS_FACTOR);
 }
 
-static zend_always_inline void gc_ref_set_info_ex(zend_refcounted *ref, gc_root_buffer *buffer, uint32_t color)
+static zend_always_inline gc_root_buffer* gc_decompress(zend_refcounted *ref, uint32_t idx)
 {
-	if (sizeof(gc_root_buffer) == 16) {
-		GC_TYPE_INFO(ref) =
-			(((char*)buffer + ((color) << 4) - (char*)GC_G(buf)) << (GC_INFO_SHIFT - 4))
-			| (GC_TYPE_INFO(ref) & (GC_TYPE_MASK | GC_FLAGS_MASK));
-	} else if (sizeof(gc_root_buffer) == 32) {
-		GC_TYPE_INFO(ref) =
-			(((char*)buffer + ((color) << 5) - (char*)GC_G(buf)) << (GC_INFO_SHIFT - 5))
-			| (GC_TYPE_INFO(ref) & (GC_TYPE_MASK | GC_FLAGS_MASK));
-	} else {
-		GC_TYPE_INFO(ref) = ((buffer - GC_G(buf)) << GC_INFO_SHIFT)
-			| (color << GC_INFO_SHIFT)
-			| (GC_TYPE_INFO(ref) & (GC_TYPE_MASK | GC_FLAGS_MASK));
+	ZEND_ASSERT(GC_NEED_COMPRESSION(idx));
+	while (idx < GC_G(first_unused)) {
+		gc_root_buffer *root = GC_G(buf) + idx;
+
+		if (root->ref == ref) {
```

Comment:
You're right, I missed the `GC_GET_PTR()`. I guess we need it in this line as well, it's just rarer to hit here.

Reply:
I've found a compromise, that keep good performance an makes small overhead only for apps with big data sets - (GC_G(first_unused) >= GC_MAX_UNCOMPRESSED).